README.md
# ProtonVPN NM Library

Copyright (c) 2021 Proton Technologies AG

This repository holds the ProtonVPN NM Library.
For licensing information see [COPYING](COPYING.md).
For contribution policy see [CONTRIBUTING](CONTRIBUTING.md).

## Description
The [ProtonVPN](https://protonvpn.com) NM (NetworkManager) Library is intended for every ProtonVPN service user.

You can download the latest stable release, either from our official repositories or directly on the [official GitHub repository](https://github.com/ProtonVPN/protonvpn-nm-lib/releases/latest).

### Dependencies:

| **Distro**                              | **Command**                                                                                                     |
|:----------------------------------------|:----------------------------------------------------------------------------------------------------------------|
|Fedora/RHEL                       | `sudo dnf install -y NetworkManager NetworkManager-openvpn NetworkManager-openvpn-gnome openvpn python3-pyxdg python3-keyring python3-jinja2 python3-distro python3-gobject libsecret-devel dbus-x11 gnome-keyring` |
|Ubuntu/Linux Mint/Debian and derivatives | `sudo apt install -y network-manager network-manager-openvpn openvpn python3-xdg python3-keyring python3-jinja2 python3-distro gir1.2-nm-1.0 libsecret-tools dbus-x11 gnome-keyring network-manager-openvpn-gnome` |
|Arch Linux/Manjaro                       | `sudo pacman -S networkmanager networkmanager-openvpn openvpn python-pyxdg python-keyring python-jinja python-distro dbus-x11 gnome-keyring` |

### Additional dependency:
[Proton Python Client](https://github.com/ProtonMail/proton-python-client)

### Virtual Environment Dependencies:
If you would like to run the the CLI from within a virtual environment (for either development purposes or other), then you can easily do that with the help of <a href="https://pipenv.readthedocs.io/en/latest/">pipenv</a>. Make sure to install pipenv and additional packages before.

| **Distro**                              | **Command**                                                                                                     |
|:----------------------------------------|:----------------------------------------------------------------------------------------------------------------|
|Fedora/RHEL                       | `sudo dnf install pkgconf-pkg-config networkmanager networkmanager-openvpn openvpn cairo-devel cairo-gobject-devel libsecret-devel gobject-introspection-devel dbus-x11 gnome-keyring` |
|Ubuntu/Linux Mint/Debian and derivatives | `sudo apt install -y pkg-config network-manager network-manager-openvpn openvpn libcairo2-dev libgirepository1.0-dev gir1.2-nm-1.0 dbus-x11 libsecret-tools gnome-keyring` |
|Arch Linux/Manjaro                       | `sudo pacman -S pkgconf networkmanager networkmanager-openvpn openvpn cairo base-devel gobject-introspection pkgconf dbus-x11 libsecret gnome-keyring gtk3` |


#### Install inside virtual environment:

  1. `cd protonvpn-nm-core`
  2. `pipenv install` (installs virtual environment and all necessary dependencies from Pipfile).
  3. `pipenv shell` (enter virtual environment).
  4. `pip install -e .` (to install).

<br>

# How to use:

### Import
``` from protonvpn_nm_lib.api import protonvpn ```

<br>

### Login
``` protonvpn.login("protonvpn@protonmail.com", "ProtonPassword") ```

<br>

### Logout
``` protonvpn.logout() ```

<br>

### Connect
``` protonvpn.setup_connection(ConnectionTypeEnum.SERVERNAME, "PT#12", ProtocolEnum.TCP) ``` <br>
``` protonvpn.connect() ``` 

<br>

### Disconnect
``` protonvpn.disconnect() ```

<br>

| **API**                              | **Description**                                                                                                     |
|:------------------------------------------------|:----------------------------------------------------------------------------------------------------------------|
| `protonvpn.login(username, password)` | Login with your Proton credentials. |
| `protonvpn.logout()` | Logout user and delete current user session. |
| `protonvpn.setup_connection(connection_type, connection_type_extra_arg, protocol)` | Setup and configure VPN connection prior to calling protonvpn.connect(). |
| `protonvpn.setup_reconnect()` | Setup and configure VPN connection to a previously connected server. Should be called before calling protonvpn.connect(). |
| `protonvpn.connect()` | Should be used either after protonvpn.setup_connection() protonvpn.setup_reconnect(). |
| `protonvpn.disconnect()` | Disconnect from ProtonVPN. |
| `protonvpn.check_session_exists()` | Check if sessions exists. |
| `protonvpn.get_connection_status()` | Get active connection status. |
| `protonvpn.get_settings()` | Get user settings. This object can be used to get and set user settings. |
| `protonvpn.get_session()` | Get user session. This object can be used to get servers list, get keyring data and other. |
| `protonvpn.get_country()` | Get country. |
| `protonvpn.get_connection_metadata()` | Get metadata of an active ProtonVPN connection. |
| `protonvpn.get_non_active_protonvpn_connection()` | Get non active ProtonVPN connection. |
| `protonvpn.get_active_protonvpn_connection()` | Get active ProtonVPN connection. |
| `protonvpn.ensure_connectivity()` | Ensure that there is an internet connection. |


debian/changelog
protonvpn-nm-lib (3.13.0-1) unstable; urgency=medium

  * Fix: Properly handle session refresh

 -- Proton Technologies AG <opensource@proton.me>  Wed, 31 Aug 2022 16:00:00 +0000
 
protonvpn-nm-lib (3.12.0-1) unstable; urgency=medium

  * Fix: Display propper error message when unable to establish VPN tunnel

 -- Proton Technologies AG <opensource@proton.me>  Fri, 29 Jul 2022 16:00:00 +0000

protonvpn-nm-lib (3.11.0-2) unstable; urgency=medium

  * Fix: When loading a session from keyring, update the app version and user agent with the current ones

 -- Proton Technologies AG <opensource@proton.me>  Tue, 31 May 2022 13:00:00 +0000
 
protonvpn-nm-lib (3.10.0-1) unstable; urgency=medium

  * Introducing Proton's refreshed look. As we continue to make privacy accessible to everyone, we've updated our apps to provide you with an even better experience with our services. Proton - Privacy by default

 -- Proton Technologies AG <opensource@proton.me>  Fri, 22 Apr 2022 13:00:00 +0000
 
protonvpn-nm-lib (3.9.0-2) unstable; urgency=medium

  * Fix: Crash when attempting to fetch netzone
  * Fix: "1000:Permission Denied" error when killswitch is enabled

 -- Proton Technologies AG <opensource@proton.me>  Wed, 20 Apr 2022 13:00:00 +0000
 
protonvpn-nm-lib (3.8.0-5) unstable; urgency=medium

  * Improve: Server list
  * Improve: Rename exception
  * Feature: Moderate NAT, Non Standard Ports

 -- Proton Technologies AG <opensource@proton.me>  Tue, 30 Nov 2021 10:00:00 +0000
 
protonvpn-nm-lib (3.7.0-6) unstable; urgency=medium

  * Improve: Handle accounting use cases

 -- Proton Technologies AG <opensource@proton.me>  Thu, 11 Nov 2021 16:30:00 +0000

protonvpn-nm-lib (3.6.1-1) unstable; urgency=medium

  * Hotfix: Update event notifications

 -- Proton Technologies AG <opensource@proton.me>  Tue, 02 Nov 2021 12:30:00 +0000

protonvpn-nm-lib (3.6.0-5) unstable; urgency=medium

  * Feature: Add notification polling
  * Improve: Human verification at login

 -- Proton Technologies AG <opensource@proton.me>  Tue, 12 Oct 2021 13:00:00 +0100
 
protonvpn-nm-lib (3.5.0-1) unstable; urgency=medium

  * Handle human verification
  * Fix: Remove network interfaces upon reboot/shutdown if kill switch is not set to permanent
  * Fix: Wake on suspend has now an additional check to ensure that the reconnector kicks in only when the session is unlocked

 -- Proton Technologies AG <opensource@proton.me>  Fri, 24 Sep 2021 11:30:00 +0100
 
protonvpn-nm-lib (3.4.1-3) unstable; urgency=medium

  * More often update server maintenance status 
  * Feature: Alternative routing 
  * Fix: Logs should be using UTC time
  * Fix: Add missing dependency

 -- Proton Technologies AG <opensource@proton.me>  Tue, 06 Jul 2021 17:45:00 +0100

protonvpn-nm-lib (3.3.2-1) unstable; urgency=medium

  * Remove IPv6 leak protection when there is no VPN and Kill Switch is disabled

 -- Proton Technologies AG <opensource@proton.me>  Wed, 23 Jun 2021 16:45:00 +0100

protonvpn-nm-lib (3.3.1-1) unstable; urgency=medium

  * Hotfix: Force disable connectivity check

 -- Proton Technologies AG <opensource@proton.me>  Tue, 22 Jun 2021 17:45:00 +0100

protonvpn-nm-lib (3.3.0-3) unstable; urgency=medium

  * Feature: Generate logs and open folder with logs

 -- Proton Technologies AG <opensource@proton.me>  Fri, 18 Jun 2021 14:30:00 +0100

protonvpn-nm-lib (3.2.2-11) unstable; urgency=medium

  * Bugfix: Fix various Kill Switch issues
  * Improve: VPN reconnector after suspend

 -- Proton Technologies AG <opensource@proton.me>  Tue, 8 Jun 2021 16:45:00 +0100

protonvpn-nm-lib (3.2.1-1) unstable; urgency=medium

  * Bugfix: Connect to re-configured Secure Core servers

 -- Proton Technologies AG <opensource@proton.me>  Fri, 4 Jun 2021 11:00:00 +0100

protonvpn-nm-lib (3.2.0-5) unstable; urgency=medium

  * Cache all necessary data after successfull authentication
  * Bugfix: Select working backend

 -- Proton Technologies AG <opensource@proton.me>  Fri, 21 May 2021 13:00:00 +0100

protonvpn-nm-lib (3.1.4-2) unstable; urgency=medium

  * Force update systemd service after upgrading the package
  * Bugfix: Filter out tor servers from connect fastest and fastest in country

 -- Proton Technologies AG <opensource@proton.me>  Wed, 11 May 2021 11:00:00 +0100

protonvpn-nm-lib (3.1.3-1) unstable; urgency=medium

  * Fix streaming services icons cache
  * Fix client config not properly getting feature flags

 -- Proton Technologies AG <opensource@proton.me>  Fri, 07 May 2021 14:30:00 +0100

 protonvpn-nm-lib (3.1.2-1) unstable; urgency=medium

  * Exclude servers with TOR feature when getting fastest server

 -- Proton Technologies AG <opensource@proton.me>  Tue, 04 May 2021 11:00:00 +0100

 protonvpn-nm-lib (3.1.1-2) unstable; urgency=medium

  * Add support for streaming and virtual locations
  * Refactor session.py so that clientconfig and streaming are own classes
  * Update python3-proton-client dependency version
  * Update incorrect dependency version for python3-proton-client

 -- Proton Technologies AG <opensource@proton.me>  Fri, 30 Apr 2021 12:00:00 +0100

 protonvpn-nm-lib (3.0.0-1) unstable; urgency=medium

  * Treat server features as bitmaps

 -- Proton Technologies AG <opensource@proton.me>  Fri, 16 Apr 2021 13:30:00 +0100

 protonvpn-nm-lib (0.5.2-1) unstable; urgency=medium

  * Cache servers and client configurations upon login

 -- Proton Technologies AG <opensource@proton.me>  Thu, 15 Apr 2021 18:30:00 +0100

 protonvpn-nm-lib (0.5.1-5) unstable; urgency=medium

  * Improve reconnection logic when computer goes to sleep or there is no internet connectivity
  * Improve logging
  * Impove Kill Switch --on option after reboot
  * Improve error handling
  * Disconnect after logout
  * Return server object after successfully reconnecting
  * Add secure core settings for GUI purpose
  * Rename Kill Switch always-on to permanent
  * Add option to connect to fastest server and fastest server in country based on secure core setting


protonvpn_nm_lib/enums.py
from enum import Enum, IntFlag


class APIEndpointEnum(Enum):
    LOGICALS = "/vpn/logicals"
    LOADS = "/vpn/loads"
    CLIENT_CONFIG = "/vpn/clientconfig"
    STREAMING_SERVICES = "/vpn/streamingservices"
    NOTIFICATIONS = "/core/v4/notifications"
    LOCATION = "/vpn/location"
    SESSIONS = "/vpn/sessions"


class NotificationEnum(Enum):
    EMPTY = "empty"
    GENERIC = "generic"


class NotificationStatusEnum(Enum):
    OPENED = 0
    NOT_OPENED = 1
    UNKNOWN = 2


class ProtocolEnum(Enum):
    TCP = "tcp"
    UDP = "udp"
    IKEV2 = "ikev2"
    WIREGUARD = "wireguard"


class ProtocolImplementationEnum(Enum):
    OPENVPN = "openvpn"
    STRONGSWAN = "strongswan"
    WIREGUARD = "wireguard"


class ProtocolPortEnum(Enum):
    TCP = 443
    UDP = 1194


class FeatureEnum(IntFlag):
    NORMAL = 0            # 0
    SECURE_CORE = 1 << 0  # 1
    TOR = 1 << 1          # 2
    P2P = 1 << 2          # 4
    STREAMING = 1 << 3    # 8
    IPv6 = 1 << 4         # 16

    @classmethod
    def list(cls):
        return list(map(lambda feature: feature, cls))


class ServerTierEnum(Enum):
    FREE = 0
    BASIC = 1
    # PLUS = 2
    # VISIONARY = 2
    PLUS_VISIONARY = 2
    PM = 3


class UserTierEnum(Enum):
    FREE = 0
    BASIC = 1
    PLUS_VISIONARY = 2
    PM = 3


class ConnectionMetadataEnum(Enum):
    SERVER = "connected_server"
    CONNECTED_TIME = "connected_time"
    PROTOCOL = "connected_protocol"
    DISPLAY_SERVER_IP = "display_server_ip"


class LastConnectionMetadataEnum(Enum):
    SERVER = ConnectionMetadataEnum.SERVER.value
    PROTOCOL = ConnectionMetadataEnum.PROTOCOL.value
    SERVER_IP = "last_connect_ip"
    DISPLAY_SERVER_IP = ConnectionMetadataEnum.DISPLAY_SERVER_IP.value


class NetzoneMetadataEnum(Enum):
    ADDRESS = "address"


class ClientSuffixEnum(Enum):
    PLATFORM = "pl"
    NETSHIELD = "f1"
    NETSHIELD_ADS_TRACKING = "f2"
    NETSHIELD_NA = "f3"
    RANDOMAZIED_NAT = "nr"


class KeyringEnum(Enum):
    DEFAULT_KEYRING_SERVICE = "ProtonVPN"
    DEFAULT_KEYRING_SESSIONDATA = "SessionData"
    DEFAULT_KEYRING_USERDATA = "UserData"
    DEFAULT_KEYRING_PROTON_USER = "ProtonUser"


class UserSettingStatusEnum(Enum):
    DISABLED = 0
    ENABLED = 1
    CUSTOM = 2


class KillswitchStatusEnum(Enum):
    DISABLED = 0
    HARD = 1
    SOFT = 2


class NetshieldStatusEnum(Enum):
    DISABLED = "f0"
    MALWARE = ClientSuffixEnum.NETSHIELD.value
    ADS_MALWARE = ClientSuffixEnum.NETSHIELD_ADS_TRACKING.value


class SecureCoreStatusEnum(Enum):
    OFF = 0
    ON = 1


class NetshieldTranslationEnum(Enum):
    DISABLED = 0
    MALWARE = 1
    ADS_MALWARE = 2


class UserSettingConnectionEnum(Enum):
    DEFAULT_PROTOCOL = "default_protocol"
    KILLSWITCH = "killswitch"
    DNS = "dns"
    DNS_STATUS = "dns_status"
    CUSTOM_DNS = "custom_dns"
    SPLIT_TUNNELING = "split_tunneling"
    SPLIT_TUNNELING_STATUS = "split_tunneling_status"
    IP_LIST = "ip_list"
    NETSHIELD = "netshield"
    SECURE_CORE = "secure_core"
    VPN_ACCELERATOR = "vpn_accelerator"
    ALTERNATIVE_ROUTING = "alternative_routing"
    EVENT_NOTIFICATION = "event_notification"
    MODERATE_NAT = "moderate_nat"
    NON_STANDARD_PORTS = "non_standard_ports"


class ProtonSessionAPIMethodEnum(Enum):
    API_REQUEST = "api_request"
    AUTHENTICATE = "authenticate"
    LOGOUT = "logout"
    FULL_CACHE = "logicals"
    LOADS_CACHE = "loads"


class MetadataActionEnum(Enum):
    GET = "get"
    WRITE = "write"
    REMOVE = "remove"


class MetadataEnum(Enum):
    CONNECTION = "connection_metadata"
    LAST_CONNECTION = "last_connection_metadata"
    SERVER_CACHE = "cache_metadata"
    NETZONE = "netzone"


class ConnectionTypeEnum(Enum):
    FREE = 0
    SERVERNAME = 1
    FASTEST = 2
    RANDOM = 3
    COUNTRY = 4
    SECURE_CORE = 5
    PEER2PEER = 6
    TOR = 7


class NetworkManagerConnectionTypeEnum(Enum):
    ACTIVE = 0
    ALL = 1


class ConnectionStatusEnum(Enum):
    SERVER_INFORMATION = "server_information"
    PROTOCOL = "protocol"
    TIME = "time"
    KILLSWITCH = "killswitch"
    NETSHIELD = "netshield"
    SERVER_IP = "server_ip"


class DisplayUserSettingsEnum(Enum):
    PROTOCOL = 0
    KILLSWITCH = 1
    DNS = 2
    CUSTOM_DNS = 3
    NETSHIELD = 4
    VPN_ACCELERATOR = 5
    ALT_ROUTING = 6
    MODERATE_NAT = 7
    NON_STANDARD_PORTS = 8


class KillSwitchInterfaceTrackerEnum(Enum):
    EXISTS = 0
    IS_RUNNING = 1


class KillSwitchActionEnum(Enum):
    PRE_CONNECTION = "pre_connection",
    POST_CONNECTION = "post_connection",
    SOFT = "soft_connection"
    ENABLE = "enable"
    DISABLE = "disable"


class DaemonReconnectorEnum(Enum):
    STOP = "stop"
    START = "start"
    DAEMON_RELOAD = "daemon-reload"


class ConnectionStartStatusEnum(Enum):
    STATE = "state"
    REASON = "reason"
    MESSAGE = "message"


class VPNConnectionStateEnum(Enum):
    """
    NMVpnConnectionState(int)

    0 (UNKNOWN): The state of the VPN connection is unknown.
    1 (PREPARING_TO_CONNECT): The VPN connection is preparing to connect.
    2 (NEEDS_CREDENTIALS): The VPN connection needs authorization credentials.
    3 (BEING_ESTABLISHED): The VPN connection is being established.
    4 (GETTING_IP_ADDRESS): The VPN connection is getting an IP address.
    5 (IS_ACTIVE): The VPN connection is active.
    6 (FAILED): The VPN connection failed.
    7 (DISCONNECTED): The VPN connection is disconnected.
    999 (UNKNOWN_ERROR): Custom error
    """
    UNKNOWN = 0
    PREPARING_TO_CONNECT = 1
    NEEDS_CREDENTIALS = 2
    BEING_ESTABLISHED = 3
    GETTING_IP_ADDRESS = 4
    IS_ACTIVE = 5
    FAILED = 6
    DISCONNECTED = 7
    UNKNOWN_ERROR = 999


class VPNConnectionReasonEnum(Enum):
    """
    NMActiveConnectionStateReason(int)

    0 (UNKNOWN):  The reason for the active connection state change
            is unknown.
    1 (NOT_PROVIDED):  No reason was given for the
        active connection state change.
    2 (USER_HAS_DISCONNECTED):  The active connection changed state because
        the user disconnected it.
    3 (DEVICE_WAS_DISCONNECTED):  The active connection changed state because
        the device it was using was disconnected.
    4 (SERVICE_PROVIDER_WAS_STOPPED):  The service providing the
        VPN connection was stopped.
    5 (IP_CONFIG_WAS_INVALID):  The IP config of the active
        connection was invalid.
    6 (CONN_ATTEMPT_TO_SERVICE_TIMED_OUT):  The connection attempt
        to the VPN service timed out.
    7 (TIMEOUT_WHILE_STARTING_VPN_SERVICE_PROVIDER):  A timeout occurred while
        starting the service providing the VPN connection.
    8 (START_SERVICE_VPN_CONN_SERVICE_FAILED):  Starting the service
        providing the VPN connection failed.
    9 (SECRETS_WERE_NOT_PROVIDED):  Necessary secrets for the connection
        were not provided.
    10 (SERVER_AUTH_FAILED): Authentication to the server failed.
    11 (DELETED_FROM_SETTINGS): The connection was deleted from settings.
    12 (MASTER_CONN_FAILED_TO_ACTIVATE): Master connection of this
        connection failed to activate.
    13 (CREATE_SOFTWARE_DEVICE_LINK_FAILED): Could not create
        the software device link.
    14 (VPN_DEVICE_DISAPPEARED): The device this connection
        depended on disappeared.
    999 (UNKNOWN_ERROR): Custom error
    """
    UNKNOWN = 0
    NOT_PROVIDED = 1
    USER_HAS_DISCONNECTED = 2
    DEVICE_WAS_DISCONNECTED = 3
    SERVICE_PROVIDER_WAS_STOPPED = 4
    IP_CONFIG_WAS_INVALID = 5
    CONN_ATTEMPT_TO_SERVICE_TIMED_OUT = 6
    TIMEOUT_WHILE_STARTING_VPN_SERVICE_PROVIDER = 7
    START_SERVICE_VPN_CONN_SERVICE_FAILED = 8
    SECRETS_WERE_NOT_PROVIDED = 9
    SERVER_AUTH_FAILED = 10
    DELETED_FROM_SETTINGS = 11
    MASTER_CONN_FAILED_TO_ACTIVATE = 12
    CREATE_SOFTWARE_DEVICE_LINK_FAILED = 13
    VPN_DEVICE_DISAPPEARED = 14
    UNKNOWN_ERROR = 999


class ServerStatusEnum(Enum):
    ACTIVE = 1
    UNDER_MAINTENANCE = 0


# ## Dbus/Systemd related
class SystemBusLogin1ObjectPathEnum(Enum):
    LOGIN1 = "/org/freedesktop/login1"
    USER_SELF = "/org/freedesktop/login1/user/self"


class SystemBusLogin1InterfaceEnum(Enum):
    LOGIN1_USER = "org.freedesktop.login1.User"
    SESSION = "org.freedesktop.login1.Session"
    MANAGER = "org.freedesktop.login1.Manager"


class SystemBusNMObjectPathEnum(Enum):
    NETWORK_MANAGER = "/org/freedesktop/NetworkManager"
    NM_SETTINGS = "/org/freedesktop/NetworkManager/Settings"


class SystemBusNMInterfaceEnum(Enum):
    NETWORK_MANAGER = "org.freedesktop.NetworkManager"
    NM_CONNECTION_SETTINGS = "org.freedesktop.NetworkManager.Settings.Connection"
    NM_SETTINGS = "org.freedesktop.NetworkManager.Settings"
    NM_CONNECTION_ACTIVE = "org.freedesktop.NetworkManager.Connection.Active"
    NM_DEVICE = "org.freedesktop.NetworkManager.Device"
